### PR TITLE
Change TestGrid structure to comply with Jenkins parallel pipeline execution

### LINKED ISSUE #971: Can not run parallel executors of the same job in multiple nodes
**Description:**
In TestGrid jobs we have a parallel block where the job is forked into multiple executors and then again joined after the parallel block. (We use the parallel block to execute job for each infra-combination)
Handing the parallel block to executors and other steps needed to run the parallel block are done by Jenkins. 
When the parallel blocks are running, in the memory, jenkins consider them has separate new jobs and create virtual directories (Ex: if the job name is: wso2is-intg, then for 3 parallel blocks, it will create wso2is-intg@2, @3 virtual directories). However though there are virtual directories, it seems all of them are pointing to the same workspace in the base directory. (As we don't have actual @2, @3 directories created in the disks).
This works well when we have only one node with multiple executors. But when there are more than one node, the other nodes don't have the base directory inside them. So if an executor in such node receives a parallel block to be executed, it brings an error as workspace cannot be found.

**Suggested Labels:**
Bug

**Suggested Assignees:**

**Affected Product Version:**

**OS, DB, other environment details and versions:**    

**Steps to reproduce:**
Execute a TestGrid job such that the parallel blocks are shared among multiple slaves.

**Related Issues:**

### PR DESCRIPTION
**Pull request by pasindujw**
**Purpose**
Fixes #971 

**Goals**
Spread parallel execution into multiple slaves.

**Approach**
**Isolate the parallel block working directories, so that each block run its own**
With this PR, the jobs will run as;
Sample directory;

_Node1_
```
testgridhome -> jobs ->  wso2is-intg -> workspace  (workspace used by generate-testplan command)
                                     -> 2 -> workspace   (workspaces used to run parallel block)
                                     -> 3 -> workspace
                                     -> 4 -> workspace    
                                     -> 5 -> workspace
```
_Node2_
```
testgridhome -> jobs ->  wso2is-intg -> 6 -> workspace   (workspaces used to run parallel block)
                                      -> 7 -> workspace
                                      -> 8 -> workspace
                                      -> 9 -> workspace
                                      -> 10 -> workspace
                                      -> 11 -> workspace
```

The PR include necessary changes to existing Job execution flow in order to make it self-contained.

**Security checks**
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: common/src/main/java/org/wso2/testgrid/common/util/S3StorageUtil.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    /**
+     * Returns the path of the test-run log file in S3 bucket.
+     * <p>
+     * /builds/#depl_name#_#infra-uuid#_#test-run-num#/test-run.log
+     *
+     * @param testPlan test-plan
+     * @param truncated whether the truncated log or the raw log file
+     * @return log file path
+     */
+    public static String deriveTestRunLogFilePath(TestPlan testPlan, Boolean truncated) {
```
Comment (kasunbg) [inline common/src/main/java/org/wso2/testgrid/common/util/S3StorageUtil.java]: #deriveTestRunLogFilePath method does not belong in a class named S3Util imo..

File: core/src/main/java/org/wso2/testgrid/core/command/RunTestPlanCommand.java [hunk 2]
```diff
@@ -131,6 +142,48 @@ public void execute() throws CommandExecutionException {
...
+     * This method resolves the absolute path to a path mentioned in
+     * the iobConfigYaml.
+     *
+     * @param path          The path to resolve
+     * @param testPlan the testPlan.yaml bean
+     * @return the resolved path
+     * @throws TestGridRuntimeException if the resolved path does not exist.
+     */
+    private String resolvePath(Path path, TestPlan testPlan) {
+        path = path.toAbsolutePath().normalize();
```
Comment (kasunbg) [inline core/src/main/java/org/wso2/testgrid/core/command/RunTestPlanCommand.java]: testPlan is an unused variable (except for the debug log).

File: core/src/test/java/org/wso2/testgrid/core/command/RunTestPlanCommandTest.java [hunk 3]
```diff
@@ -130,6 +135,11 @@ public void init() throws Exception {
         testPlan.setDeploymentPattern(deploymentPattern);
         testPlan.setInfraParameters(infraParamsString);
         testPlan.setDeployerType(TestPlan.DeployerType.SHELL);
+        testPlan.setScenarioTestsRepository(workspaceDir + "/workspace/scenarioTests");
```
Comment (kasunbg) [inline core/src/test/java/org/wso2/testgrid/core/command/RunTestPlanCommandTest.java]: We need to use Paths API or the File.separator.

File: core/src/test/java/org/wso2/testgrid/core/command/RunTestPlanCommandTest.java [hunk 4]
```diff
@@ -167,6 +177,12 @@ public void testExecute(String jobConfigFile) throws Exception {
         Assert.assertEquals(content, "Deploy server1\n", "my-deployment.txt file does not contain expected content.");
     }
 
+    private void copyWorkspaceArtifacts() throws IOException {
+        String resourceDir = "./src/test/resources/workspace";
+            FileUtils.copyDirectory(new File(resourceDir), new File(workspaceDir + "/workspace"));
+            logger.info("Copying necessary artifacts to workspace " + workspaceDir + "/workspace");
```
Comment (kasunbg) [inline core/src/test/java/org/wso2/testgrid/core/command/RunTestPlanCommandTest.java]: Formatting issue. We need to think about File system differences in Windows and Linux.

File: reporting/src/main/java/org/wso2/testgrid/reporting/TestReportEngine.java [hunk 5]
```diff
@@ -825,6 +825,7 @@ private void postProcessSummaryTable(Map<String, InfrastructureBuildStatus> test
                 Optional<TestPlan> testPlanById = testPlanUOW.getTestPlanById(testPlanYaml.getId());
                 if (testPlanById.isPresent()) {
                     TestPlan mergedTestPlan = TestGridUtil.mergeTestPlans(testPlanYaml, testPlanById.get(), false);
+                    mergedTestPlan.setWorkspace(workspace);
```
Comment (kasunbg) [inline reporting/src/main/java/org/wso2/testgrid/reporting/TestReportEngine.java]: Setting workspace like this in a unrelated class like TestReportEngine really complicate the matters in the long run! Shall we see a better way to fix this?

File: common/src/main/java/org/wso2/testgrid/common/util/S3StorageUtil.java (hunk 1 above)
Comment (pasindujw) [reply common/src/main/java/org/wso2/testgrid/common/util/S3StorageUtil.java]: This is referring to getting the TestRunLogFile Path(URL) in S3 bucket. I hope it is fine if the method name and method comments are changed to reflect that :) WDYT?

PS: We have the same method in TestGridUtil (Where the path is generated referring to the workspace directory) which is used when using the local storage.

File: reporting/src/main/java/org/wso2/testgrid/reporting/TestReportEngine.java (hunk 5 above)
Comment (pasindujw) [reply reporting/src/main/java/org/wso2/testgrid/reporting/TestReportEngine.java]: When we create a TestPlan object in the middle of the flow by calling the database or so, it is missing the workspace property. So in such times the rule would be; **whenever a TestPlan object is created, right then we need to set its workspace (which we receive at runtime via input argument)**. 
Agreed it complicates the process :) 
The ideal solution would be to keep the `workspace` in a Context where we keep run-time details of the Job. :)

File: common/src/main/java/org/wso2/testgrid/common/util/S3StorageUtil.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
+package org.wso2.testgrid.common.util;
+
+import org.wso2.testgrid.common.TestGridConstants;
+import org.wso2.testgrid.common.TestPlan;
+import org.wso2.testgrid.common.config.ConfigurationContext;
+
+import java.nio.file.Paths;
+
+/**
+ * This Util class holds the utility methods used to manage TestGrid S3 storage .
```
Comment (harshanL) [inline common/src/main/java/org/wso2/testgrid/common/util/S3StorageUtil.java]: Unnecessary empty space before '.'.

File: core/src/main/java/org/wso2/testgrid/core/command/RunTestPlanCommand.java [hunk 7]
```diff
@@ -131,6 +142,44 @@ public void execute() throws CommandExecutionException {
...
+        //keyfile
+        if (testPlan.getKeyFileLocation() != null) {
+            repoPath = Paths.get(testPlan.getWorkspace(), testPlan.getKeyFileLocation());
+            testPlan.setKeyFileLocation(resolvePath(repoPath));
+        }
+    }
+
+    /**
+     * This method resolves the absolute path to a path mentioned in
+     * the iobConfigYaml.
```
Comment (harshanL) [inline core/src/main/java/org/wso2/testgrid/core/command/RunTestPlanCommand.java]: Misspelled 'jobConfigYaml'.

File: core/src/main/java/org/wso2/testgrid/core/command/RunTestPlanCommand.java [hunk 8]
```diff
@@ -131,6 +142,44 @@ public void execute() throws CommandExecutionException {
...
+     * @throws TestGridRuntimeException if the resolved path does not exist.
+     */
+    private String resolvePath(Path path) {
+        path = path.toAbsolutePath().normalize();
+        if (!Files.exists(path)) {
+            throw new TestGridRuntimeException("Path '" + path.toString() + "' does not exist.");
+        }
+        return path.toString();
+    }
     /**
```
Comment (harshanL) [inline core/src/main/java/org/wso2/testgrid/core/command/RunTestPlanCommand.java]: Add an empty line here.

File: core/src/test/java/org/wso2/testgrid/core/command/RunTestPlanCommandTest.java [hunk 9]
```diff
@@ -107,14 +110,16 @@
     private TestPlanExecutor testPlanExecutor;
     private ScenarioExecutor scenarioExecutor;
     private String actualTestPlanFileLocation;
-
+    private String workspaceDir;
     @BeforeMethod
```
Comment (harshanL) [inline core/src/test/java/org/wso2/testgrid/core/command/RunTestPlanCommandTest.java]: Add an empty line here.

File: core/src/test/java/org/wso2/testgrid/core/command/RunTestPlanCommandTest.java [hunk 10]
```diff
@@ -167,6 +177,13 @@ public void testExecute(String jobConfigFile) throws Exception {
         Assert.assertEquals(content, "Deploy server1\n", "my-deployment.txt file does not contain expected content.");
     }
 
+    private void copyWorkspaceArtifacts() throws IOException {
+        String resourceDir = Paths.get("./src", "test", "resources", "workspace").toString();
+        String destinationDir = Paths.get(workspaceDir, "/workspace").toString();
+            FileUtils.copyDirectory(new File(resourceDir), new File(destinationDir));
+            logger.info("Copying necessary artifacts to directory: " + destinationDir);
+    }
```
Comment (harshanL) [inline core/src/test/java/org/wso2/testgrid/core/command/RunTestPlanCommandTest.java]: Please reformat the above two lines.

**Pull request merged**